component/pprof/scrape/manager.go

```diff
@@ -0,0 +1,199 @@
+package scrape
+
+import (
+	"errors"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+
+	"github.com/grafana/agent/component/pprof"
+)
+
+var reloadInterval = 5 * time.Second
+
+type Manager struct {
+	logger log.Logger
+
+	graceShut  chan struct{}
+	appendable pprof.Appendable
+
+	mtxScrape     sync.Mutex // Guards the fields below.
+	config        Arguments
+	targetsGroups map[string]*scrapePool
+	targetSets    map[string][]*targetgroup.Group
+
+	triggerReload chan struct{}
+}
+
+func NewManager(appendable pprof.Appendable, logger log.Logger) *Manager {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	return &Manager{
+		logger:        logger,
+		appendable:    appendable,
+		graceShut:     make(chan struct{}),
+		triggerReload: make(chan struct{}, 1),
+		targetsGroups: make(map[string]*scrapePool),
+		targetSets:    make(map[string][]*targetgroup.Group),
+	}
+}
+
+// Run receives and saves target set updates and triggers the scraping loops reloading.
+// Reloading happens in the background so that it doesn't block receiving targets updates.
+func (m *Manager) Run(tsets <-chan map[string][]*targetgroup.Group) {
+	go m.reloader()
+	for {
+		select {
+		case ts := <-tsets:
+			m.updateTsets(ts)
+
+			select {
+			case m.triggerReload <- struct{}{}:
+			default:
+			}
+
+		case <-m.graceShut:
+			return
+		}
+	}
+}
+
+func (m *Manager) reloader() {
+	ticker := time.NewTicker(reloadInterval)
+
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.graceShut:
+			return
+		case <-ticker.C:
+			select {
+			case <-m.triggerReload:
+				m.reload()
+			case <-m.graceShut:
+				return
+			}
+		}
+	}
+}
+
+func (m *Manager) reload() {
+	m.mtxScrape.Lock()
+	var wg sync.WaitGroup
+	for setName, groups := range m.targetSets {
+		if _, ok := m.targetsGroups[setName]; !ok {
+			sp, err := newScrapePool(m.config, m.appendable, log.With(m.logger, "scrape_pool", setName))
+			if err != nil {
+				level.Error(m.logger).Log("msg", "error creating new scrape pool", "err", err, "scrape_pool", setName)
+				continue
+			}
+			m.targetsGroups[setName] = sp
+		}
+
+		wg.Add(1)
+		// Run the sync in parallel as these take a while and at high load can't catch up.
+		go func(sp *scrapePool, groups []*targetgroup.Group) {
+			sp.sync(groups)
+			wg.Done()
+		}(m.targetsGroups[setName], groups)
+	}
+	m.mtxScrape.Unlock()
+	wg.Wait()
+}
+
+// ApplyConfig resets the manager's target providers and job configurations as defined by the new cfg.
+func (m *Manager) ApplyConfig(cfg Arguments) error {
+	m.mtxScrape.Lock()
+	defer m.mtxScrape.Unlock()
+	if reflect.DeepEqual(m.config, cfg) {
```

Comment:
Same as Robert's comment above, that Update will only be called with the updated Arguments, even if it's in a different package.